app/controllers/businessactivities/WhoIsYourAccountantUkAddressController.scala

```diff
@@ -0,0 +1,76 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.businessactivities
+
+import com.google.inject.Inject
+import connectors.DataCacheConnector
+import controllers.DefaultBaseController
+import forms.{EmptyForm, Form2, InvalidForm, ValidForm}
+import models.businessactivities.{AccountantsAddress, BusinessActivities, UkAccountantsAddress}
+import services.AutoCompleteService
+import uk.gov.hmrc.play.frontend.auth.connectors.AuthConnector
+import utils.AuthAction
+
+class WhoIsYourAccountantUkAddressController @Inject()(val dataCacheConnector: DataCacheConnector,
+                                                       val autoCompleteService: AutoCompleteService,
+                                                       val authAction: AuthAction
+                                              )extends DefaultBaseController {
+
+  private val defaultValues:AccountantsAddress = UkAccountantsAddress("", "", None, None, "")
+
+  def get(edit: Boolean = false) = authAction.async {
+    implicit request =>
+      dataCacheConnector.fetch[BusinessActivities](request.credId, BusinessActivities.key) map {
+        response =>
+          val form = (for {
+            businessActivities <- response
+            whoIsYourAccountant <- businessActivities.whoIsYourAccountant.flatMap(acc => acc.address)
+          } yield {
+            if(whoIsYourAccountant.isUk) { Form2[AccountantsAddress](whoIsYourAccountant) } else { EmptyForm }
+          }).getOrElse(Form2(defaultValues))
+          Ok(views.html.businessactivities.who_is_your_accountant_uk_address(
+            form, edit,
+            response.flatMap(ba => ba.whoIsYourAccountant).flatMap(acc => acc.names).map(names => names.accountantsName).getOrElse("!!")))
+      }
+  }
+
+  def post(edit : Boolean = false) = authAction.async {
+    implicit request =>
+      Form2[UkAccountantsAddress](request.body) match {
+        case f: InvalidForm =>
+          dataCacheConnector.fetch[BusinessActivities](request.credId, BusinessActivities.key) map {
+            response => BadRequest(views.html.businessactivities.who_is_your_accountant_uk_address(f, edit, getName(response)))
+          }
+        case ValidForm(_, data) => {
+          for {
+            businessActivity <- dataCacheConnector.fetch[BusinessActivities](request.credId, BusinessActivities.key)
+            _ <- dataCacheConnector.save[BusinessActivities](request.credId, BusinessActivities.key,
+              businessActivity.whoIsYourAccountant(businessActivity.flatMap(ba => ba.whoIsYourAccountant).map(acc => acc.copy(address = Option(data))))
+            )
+          } yield if (edit) {
+            Redirect(routes.SummaryController.get())
+          } else {
+            Redirect(routes.TaxMattersController.get())
+          }
+        }
+      }
+  }
+
+  private def getName(response: Option[BusinessActivities]): String = response
```

Comment:
As above. Shared collection of simplified methods?

Reply:
As above